### PR TITLE
[npm node] support for custom npm registry

### PR DESCRIPTION
**Pull request by iamstarkov**
there are two types of custom npm registries:
* public: due to geographical or maybe political reasons https://registry.npm.taobao.org/—essentially copy of https://registry.npmjs.org/
* private: due to legal requirements (my case) or redundancy reasoning—either npm enterprise or sinopia

affected endpoints:
* _npm node version integration_ now supports `/npm/https/registry.npm.taobao.org/v/express.svg`:
  ```diff
  -/^\/npm\/v\/(?:@([^/]+))?\/?([^/]*)\/?([^/]*)\.(svg|png|gif|jpg|json)$/
  +/^\/npm(?:\/(http(?:s)?)\/([^/]+))?\/v\/(?:@([^/]+))?\/?([^/]*)\/?([^/]*)\.(svg|png|gif|jpg|json)$/
  ```
* _npm license integration_ now supports `/npm/https/registry.npm.taobao.org/l/express.svg`:
  ```diff
  -/^\/npm\/l\/(?:@([^/]+)\/)?([^/]+)\.(svg|png|gif|jpg|json)$/
  +/^\/npm(?:\/(http(?:s)?)\/([^/]+))?\/l\/(?:@([^/]+)\/)?([^/]+)\.(svg|png|gif|jpg|json)$/
  ```
* _npm node version integration_ now supports `/node/https/registry.npm.taobao.org/v/express.svg`:
  ```diff
  -/^\/npm\/l\/(?:@([^/]+)\/)?([^/]+)\.(svg|png|gif|jpg|json)$/
  +/^\/npm(?:\/(http(?:s)?)\/([^/]+))?\/l\/(?:@([^/]+)\/)?([^/]+)\.(svg|png|gif|jpg|json)$/
  ```

downloads related endpoints are not affected due to custom https://api.npmjs.org/ server, which complicates stuff. But apart from complexity, there are several reasons not to modify these endpoints:
  * for public registries its (maybe) okay to use default registry
  * for private registries this downloads metric is (maybe) unnecessary

P.S. inspiration for `/npm(/:protocol/:host)/…` is jenkins endpoint

### CODE REVIEW

**Comment by paulmelnikow** [comment]
Thanks for this!

I appreciate that you're modeling it the Jenkins badge. Since #1186 it's possible to pass arbitrary parameters through the query string. I'd like to use this for exception / rarely used options, long options, or options which require munging to fit into the URL. This seems like a good place for it. Basically, something like `?registry_uri=`.

The benefit is a more consistent and understandable URL pattern. The resultant URL isn't quite as pretty and takes a bit of effort to encode, but it's also not going to be used by many people, and after #701 it should be really easy to encode one of these through the UI.

You can see a few examples in `server.js` where an array of query string field names is passed in. This is from the dynamic JSON badge:

```js
queryParams: ['uri', 'query', 'prefix', 'suffix'],
```

File: lib/all-badge-examples.js [hunk 1]
```diff
@@ -647,6 +647,34 @@ const allBadgeExamples = [
...
+      {
+        title: 'custom registry :: npm (tag)',
+        previewUri: '/npm/https/registry.npm.taobao.org/v/npm/next.svg',
+        keywords: [
+          'node'
+        ]
+      },
+      {
+        title: 'custom registry :: npm (scoped with tag)',
+        previewUri: '/npm/https/registry.npm.taobao.org/v/@cycle/core/canary.svg',
```
Comment (paulmelnikow) [inline lib/all-badge-examples.js]: How about choosing one of these four and removing the others? The title should be something like **npm (scoped with tag, custom registry)**. Same with the node badges. I feel like people will understand the pattern.

File: lib/npm-provider.js [hunk 2]
```diff
@@ -36,6 +36,15 @@ function mapNpmDownloads(camp, urlComponent, apiUriComponent) {
   }));
 }
 
+// npm registry url
+function getNpmRegistryUrl(maybeProtocol, maybeHost) {
```
Comment (paulmelnikow) [inline lib/npm-provider.js]: Using the query string should make this a little simpler. If you still need this function it should go in a new file `npm-badge-helpers.js`.

File: server.js [hunk 3]
```diff
@@ -1658,22 +1661,23 @@ cache(function(data, match, sendBadge, request) {
...
       // Use this endpoint as an optimization. It covers the vast majority of
       // these badges, and the response is smaller.
-      apiUrl = `https://registry.npmjs.org/${packageName}/${registryTag}`;
+      apiUrl = `${registryUrl}/${packageName}/${registryTag}`;
   } else {
     // e.g. https://registry.npmjs.org/@cedx%2Fgulp-david
     // because https://registry.npmjs.org/@cedx%2Fgulp-david/latest does not work
     const path = encodeURIComponent(`${scope}/${packageName}`);
-    apiUrl = `https://registry.npmjs.org/@${path}`;
+    apiUrl = `${registryUrl}/@${path}`;
```
Comment (paulmelnikow) [inline server.js]: I think I wrote tests for the current Node and NPM badges. Could you add tests of the new routes?

File: lib/npm-provider.js (hunk 2 above)
Comment (iamstarkov) [reply lib/npm-provider.js]: what is the difference between `npm-provider` and `npm-badge-helpers`?

File: server.js (hunk 3 above)
Comment (iamstarkov) [reply server.js]: sure

**Comment by iamstarkov** [comment]
no problems. i'll change stuff. just to be on the same page. here this is the list to change/add:

* [ ] query params instead of complicating the url pattern
* [ ] remove extra examples
* [ ] tests

**Comment by iamstarkov** [comment]
and thanks for the feedback

File: lib/npm-provider.js (hunk 2 above)
Comment (paulmelnikow) [reply lib/npm-provider.js]: Most of the badge providers are in `server.js`, though a few, like npm, are pulled out. We use `-provider` to refer to the function which implements the badge.

The `-badge-helpers` are service-specific utility functions which are invoked by the providers.

Yea, in a couple cases the line is gray. Though this one seems to be a utility function.

**Comment by paulmelnikow** [comment]
Yup, that's it. And if you keep the helper function, please move it to a different file.

Thanks!

**Comment by iamstarkov** [comment]
only tests left

**Comment by iamstarkov** [comment]
done

File: lib/all-badge-examples.js (hunk 1 above)
Comment (iamstarkov) [reply lib/all-badge-examples.js]: screenshot-2017-11-19 shields io quality metadata badges for open source projects 2
screenshot-2017-11-19 shields io quality metadata badges for open source projects 1

Comment (iamstarkov) [reply lib/all-badge-examples.js]: it looks like this now, hope its okay

File: lib/npm-provider.js (hunk 2 above)
Comment (iamstarkov) [reply lib/npm-provider.js]: moved it

File: server.js (hunk 3 above)
Comment (iamstarkov) [reply server.js]: done

File: lib/npm-provider.js (hunk 2 above)
Comment (iamstarkov) [reply lib/npm-provider.js]: and i got rid off of helper function, now i only use default npm registry uri as a constant to share

**Comment by iamstarkov** [comment]
because of added queryParams and following added indentation to handler diff looks messy.

to make things simpler to review use diff url without whitespace diff https://github.com/badges/shields/pull/1277/files?w=1

**Comment by iamstarkov** [comment]
tests are ok, its just ping to chinese registry is long and mocha is timing out. 

what do you think?

**Comment by paulmelnikow** [comment]
Excellent! Thanks for the diff; it's easy enough to read. This looks great. 

Only one thing I'd ask for, is tests of the license badge. It seems think we have any now. The reason it seems important is that all that code was in flight. You could just pick a package and hard-code the expected value.

Is there another mirror that's faster to reach from the U.S.? Alternatively we could just leave the flaky test. I've solved the timeout issue in IcedFrisby; it's waiting to be shipped in 2.0.

**Comment by iamstarkov** [comment]
sure, i'll add tests for license service

**Comment by iamstarkov** [comment]
about registries

**Comment by iamstarkov** [comment]
its complicated. there used to be some semi-supported mirrors. Nowadays its only `registry.npm.taobao.org` and `r.cnpmjs.org` with average ping 275ms and 425ms respectively. So if taobao isnt a good fit, then r.cnpmjs isnt either.

**Comment by iamstarkov** [comment]
```
~/projects/oss/shields feat/custom-npm-registry
☯ npm i -g nrm -dddddd
npm info it worked if it ends with ok
npm info using npm@5.5.1
npm info using node@v8.9.1
npm http fetch GET 304 https://registry.npmjs.org/nrm 938ms (from cache)
npm http fetch GET 304 https://registry.npmjs.org/async 96ms (from cache)
npm http fetch GET 304 https://registry.npmjs.org/ini 125ms (from cache)
npm http fetch GET 304 https://registry.npmjs.org/node-echo 125ms (from cache)
npm http fetch GET 304 https://registry.npmjs.org/extend 129ms (from cache)
npm http fetch GET 304 https://registry.npmjs.org/request 146ms (from cache)
npm http fetch GET 304 https://registry.npmjs.org/npm 228ms (from cache)
npm http fetch GET 304 https://registry.npmjs.org/only 254ms (from cache)
npm http fetch GET 304 https://registry.npmjs.org/open 784ms (from cache)
npm http fetch GET 304 https://registry.npmjs.org/commander 797ms (from cache)
npm http fetch GET 304 https://registry.npmjs.org/mkdirp 60ms (from cache)
npm http fetch GET 304 https://registry.npmjs.org/coffee-script 139ms (from cache)
npm http fetch GET 304 https://registry.npmjs.org/jistype 990ms (from cache)
```

from this log default registry is `registry.npmjs.org`. but there is as well `registry.npmjs.org` which is basically a dns alias to default one:

```
☯ ping registry.npmjs.org -c 11
PING a.sni.fastly.net (151.101.84.162): 56 data bytes
avg 48ms

☯ ping registry.npmjs.com -c 11
PING a.sni.fastly.net (151.101.84.162): 56 data bytes
avg 11 ms
```

**Comment by iamstarkov** [comment]
i'll use `registry.npmjs.com`, because its not default one, but fast

**Comment by iamstarkov** [comment]
yay, `registry.npmjs.com` fixed it

**Comment by iamstarkov** [comment]
done

**Comment by iamstarkov** [comment]
if anything left to fix, this pull-request should be fine. @paulmelnikow thanks for your input, feedback and patience

**Comment by paulmelnikow** [comment]
Thanks for all _your_ work!

**Comment by iamstarkov** [comment]
hurray

**Comment by iamstarkov** [comment]
thanks

**Comment by iamstarkov** [comment]
will you release it to https://shields.io/ anytime soon?

**Comment by paulmelnikow** [comment]
I can't; I don't have access. Deploys usually happen every 1–3 weeks. Thaddée, who has limited time on this project, is the only sysadmin. He's working on giving me access to deploy and logs, but doing so is complicated because the hosting account (and maybe the servers too) are shared with other services he runs.

**Pull request merged**